enhancement: integrate w/ watson eval

george-lhj opened the pull request:
fixes #726 I've integrated watsonx evaluation for our workflows, and have supported metrics such as eval time, status, relevance, faithfulness, context relevance (with the capability to extend to many more metrics, like harrasment, for example).

This PR only focuses on fully integrating the library into our backend. I will open further PRs related to database structure, additional capabilies for other metrics.



To test: 

1) need an ibm cloud api key: [here](https://www.ibm.com/products/watsonx-governance?utm_content=SRCWW&p1=Search&p4=2275961165693&p5=e&p9=158778237493&gclsrc=aw.ds&gad_source=1&gad_campaignid=20116653496&gbraid=0AAAAAD-_QsQR1wfDyN0j3MXkIWbCddJWu&gclid=CjwKCAjw89jGBhB0EiwA2o1On7Odp5k2DAEoYjwJf-y3jesCIAMTINqexl8AVrV3LPYzj3MuTaJ3gBoCMmgQAvD_BwE)

2) need ollama/maestro installed, and specify python version to 3.11

3) uv install the required library `ibm-watsonx-gov[agentic]`, there might be additional dependencies inside that library itself.

4) 

After creatng the agents:

export MAESTRO_AUTO_EVALUATION=true

# Mock agent
`maestro run tests/yamls/agents/evaluation_test_agent.yaml tests/yamls/workflows/evaluation_test_workflow.yaml`

# OpenAI Agent
`maestro run tests/yamls/agents/openai_agent.yaml tests/yamls/workflows/openai_workflow.yaml`



further todos:
- set up database to store these results

planetf1 reviewed (changes requested):
Two issues I noted

1. The PR text or README doesn't state what to do with the API key
2. When run, I get the following error (Python 3.11 as per info above)

```
(.venv) ➜  maestro git:(watson_eval) maestro run tests/yamls/agents/evaluation_test_agent.yaml tests/yamls/workflows/evaluation_test_workflow.yaml
Traceback (most recent call last):
  File "/Users/jonesn/src/maestro/.venv/bin/maestro", line 4, in <module>
    from maestro.cli.run_maestro import __run_cli
  File "/Users/jonesn/src/maestro/src/maestro/__init__.py", line 4, in <module>
    from .workflow import Workflow
  File "/Users/jonesn/src/maestro/src/maestro/workflow.py", line 17, in <module>
    from maestro.agents.agent_factory import AgentFramework, AgentFactory
  File "/Users/jonesn/src/maestro/src/maestro/agents/agent_factory.py", line 8, in <module>
    from .openai_agent import OpenAIAgent
  File "/Users/jonesn/src/maestro/src/maestro/agents/openai_agent.py", line 15, in <module>
    from agents import (
  File "/Users/jonesn/src/maestro/.venv/lib/python3.12/site-packages/agents/__init__.py", line 8, in <module>
    from .agent import Agent, AgentBase, ToolsToFinalOutputFunction, ToolsToFinalOutputResult
  File "/Users/jonesn/src/maestro/.venv/lib/python3.12/site-packages/agents/agent.py", line 10, in <module>
    from openai.types.responses.response_prompt_param import ResponsePromptParam
ModuleNotFoundError: No module named 'openai.types.responses.response_prompt_param'
(.venv) ➜  maestro git:(watson_eval)
```

planetf1 commented on README.md:
````diff
@@ -88,6 +88,30 @@ uv run ruff format
 ```
 
+## Agent Evaluation
+
````
Also could include model (by inspection it's llama3.1:latest on ollama)

planetf1 commented on docs/WATSONX_EVALUATION_README.md:
```diff
@@ -0,0 +1,243 @@
+- If a required input is missing, that metric is skipped gracefully.
+
+#### 3. **Environment Setup**
+- **Python 3.11**: Dedicated `.venv-eval` environment for watsonx compatibility
```
is 3.11 a testing statement? We now have 3.12 in maestro also, with 3.13/3.14 expected
(As an optional extra, if 3.11 is a hard dependency due to third parties that's fine - may be worth elaborating)

planetf1 commented on docs/WATSONX_EVALUATION_README.md:
````diff
@@ -0,0 +1,243 @@
+tests/yamls/workflows/evaluation_test_workflow.yaml # Test workflow
+```
+
+#### How It Works
````
can we expand on this with links. Without context about watsonx it's not clear what the decorators or middleware are

planetf1 commented on docs/WATSONX_EVALUATION_README.md:
```diff
@@ -0,0 +1,243 @@
+source .venv-eval/bin/activate
+
+# Install dependencies (already done)
+pip install "ibm-watsonx-gov[agentic]"
```
prefix each command with `uv` ?

planetf1 commented on docs/WATSONX_EVALUATION_README.md:
````diff
@@ -0,0 +1,243 @@
+pip install -e .
+```
+
+Note: watsonx requires Python 3.11. Keep evaluation in `.venv-eval` while core Maestro may run on 3.12+.
````
it's not clear from the instructions what is run in this eval environment?

planetf1 commented:
> 1. The PR text or README doesn't state what to do with the API key
Update - I found the text for this. apologies.

george-lhj commented on README.md:
````diff
@@ -88,6 +88,30 @@ uv run ruff format
+### Quick Start
+```bash
+# Activate evaluation environment
+source .venv-eval/bin/activate
````
I've pointed this portion to the actual watsonx eval readme

george-lhj commented on docs/WATSONX_EVALUATION_README.md:
````diff
@@ -0,0 +1,243 @@
+
+1. **Activate the evaluation environment**:
+   ```bash
+   source .venv-eval/bin/activate
````
I've pointed the README directly to this file so that they are consistent

george-lhj commented:
@planetf1 
I've finished testing and identified the issue, evaluation setup was failing with dependency conflicts between openai-agents and ibm-watsonx-gov.
Root Cause:
ibm-watsonx-gov[agentic]==1.2.2 requires openai==1.75.0 and litellm==1.67.6
openai-agents requires openai>=2.0.0 for some features.

I've updated and simplified the readme to reflect this

george-lhj commented on tests/yamls/agents/evaluation_test_agent.yaml:
```diff
@@ -0,0 +1,13 @@
+apiVersion: maestro/v1alpha1
```
I've added a basic integration test which GH actions can run, checks the evaluate shape/output, but doesn't call any of the actual evaluate function so we don't need a key
